// contentful-cli/dist/cmds/config_cmds/list.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.handler = exports.listHandler = exports.aliases = exports.builder = exports.desc = exports.command = undefined;

var _async = require('../../utils/async');

var _emojic = require('emojic');

var _emojic2 = _interopRequireDefault(_emojic);

var _text = require('../../utils/text');

var _log = require('../../utils/log');

var _proxy = require('../../utils/proxy');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const command = exports.command = 'list';

const desc = exports.desc = 'List all your configs';

const builder = exports.builder = yargs => {
  return yargs.usage('Usage: contentful config list');
};

const aliases = exports.aliases = ['ls'];
const listHandler = exports.listHandler = async ({ context }) => {
  const configList = Object.keys(context).map(key => {
    if (key === 'proxy') {
      return `${_emojic2.default.gear}  ${key}: ${(0, _proxy.proxyObjectToString)(context[key])}`;
    } else {
      return `${_emojic2.default.gear}  ${key}: ${context[key]}`;
    }
  });
  (0, _log.success)((0, _text.frame)(configList.join('\n')));
};

const handler = exports.handler = (0, _async.handleAsyncError)(listHandler);
