// contentful-migration/built/lib/intent/content-type-create.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const base_intent_1 = __importDefault(require("./base-intent"));
const content_type_create_1 = require("../action/content-type-create");
const chalk_1 = __importDefault(require("chalk"));
class ContentTypeCreateIntent extends base_intent_1.default {
    isContentTypeCreate() {
        return true;
    }
    groupsWith() {
        return false;
    }
    endsGroup() {
        return false;
    }
    toActions() {
        return [
            new content_type_create_1.ContentTypeCreateAction(this.getContentTypeId())
        ];
    }
    toPlanMessage() {
        return {
            heading: chalk_1.default `Create Content Type {bold.yellow ${this.getContentTypeId()}}`,
            details: [],
            sections: []
        };
    }
}
exports.default = ContentTypeCreateIntent;
