// contentful-migration/built/lib/intent/entryeditor-reset-to-default.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const base_intent_1 = __importDefault(require("./base-intent"));
const chalk_1 = __importDefault(require("chalk"));
const editorinterface_save_1 = require("../action/editorinterface-save");
const entryeditor_reset_to_default_1 = require("../action/entryeditor-reset-to-default");
class EntryEditorResetToDefaultIntent extends base_intent_1.default {
    isEditorInterfaceIntent() {
        return true;
    }
    isGroupable() {
        return false;
    }
    groupsWith() {
        return false;
    }
    endsGroup() {
        return false;
    }
    shouldSave() {
        return false;
    }
    shouldPublish() {
        return false;
    }
    toActions() {
        return [
            new entryeditor_reset_to_default_1.EntryEditorResetToDefaultAction(this.payload.contentTypeId),
            new editorinterface_save_1.SaveEditorInterfaceAction(this.payload.contentTypeId)
        ];
    }
    toPlanMessage() {
        return {
            heading: chalk_1.default `Reset entry editor interface for content type {bold.yellow ${this.getContentTypeId()}} to default`,
            details: [],
            sections: []
        };
    }
}
exports.default = EntryEditorResetToDefaultIntent;
