// contentful-migration/built/lib/intent/sidebarwidget-update.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const base_intent_1 = __importDefault(require("./base-intent"));
const chalk_1 = __importDefault(require("chalk"));
const editorinterface_save_1 = require("../action/editorinterface-save");
const sidebarwidget_update_1 = require("../action/sidebarwidget-update");
class SidebarWidgetUpdateIntent extends base_intent_1.default {
    isSidebarUpdate() {
        return true;
    }
    isEditorInterfaceIntent() {
        return true;
    }
    isGroupable() {
        return true;
    }
    groupsWith(other) {
        return other.isGroupable()
            && other.isEditorInterfaceIntent()
            && this.isSameContentType(other);
    }
    endsGroup() {
        return false;
    }
    shouldSave() {
        return false;
    }
    shouldPublish() {
        return false;
    }
    toActions() {
        return [
            new sidebarwidget_update_1.SidebarWidgetUpdateAction(this.payload.contentTypeId, this.payload.sidebarWidget.widgetId, this.payload.sidebarWidget.widgetNamespace, this.payload.sidebarWidget.settings, this.payload.sidebarWidget.disabled),
            new editorinterface_save_1.SaveEditorInterfaceAction(this.payload.contentTypeId)
        ];
    }
    toPlanMessage() {
        const { settings, widgetId, widgetNamespace } = this.payload.sidebarWidget;
        const settingDetails = Object.keys(settings || {}).map(settingName => chalk_1.default `{italic ${settingName}}: "${settings[settingName].toString()}"`);
        const updateDetails = [
            chalk_1.default `{italic widgetNamespace}: "${widgetNamespace}"`,
            ...settingDetails
        ];
        return {
            heading: chalk_1.default `Update editor interface for Content Type {bold.yellow ${this.getContentTypeId()}}`,
            details: [],
            sections: [{
                    heading: chalk_1.default `Update sidebar widget {yellow ${widgetId}}`,
                    details: updateDetails
                }]
        };
    }
}
exports.default = SidebarWidgetUpdateIntent;
