// contentful-cli/dist/cmds/extension_cmds/get.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.handler = exports.builder = exports.desc = exports.command = undefined;

var _contentfulClients = require('../../utils/contentful-clients');

var _async = require('../../utils/async');

var _logAsTable = require('./utils/log-as-table');

const command = exports.command = 'get';

const desc = exports.desc = 'Show an extension';

const builder = exports.builder = yargs => {
  return yargs.option('id', { type: 'string', demand: true, describe: 'Extension id' }).option('management-token', { alias: 'mt', type: 'string', describe: 'Contentful management API token' }).option('space-id', { type: 'string', describe: 'Space id' }).option('environment-id', { type: 'string', describe: 'Environment id' }).epilog('Copyright 2018 Contentful, this is a BETA release');
};

async function getExtension({ context, id }) {
  const { managementToken, activeSpaceId, activeEnvironmentId } = context;

  const client = await (0, _contentfulClients.createManagementClient)({
    accessToken: managementToken,
    feature: 'extension-get'
  });

  const space = await client.getSpace(activeSpaceId);
  const environment = await space.getEnvironment(activeEnvironmentId);
  const extension = await environment.getUiExtension(id);

  (0, _logAsTable.logExtension)(extension, activeSpaceId, activeEnvironmentId);
}

const handler = exports.handler = (0, _async.handleAsyncError)(getExtension);
