// contentful-cli/dist/cmds/organization_cmds/list.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.handler = exports.aliases = exports.builder = exports.desc = exports.command = undefined;
exports.organizationList = organizationList;

var _cliTable = require('cli-table3');

var _cliTable2 = _interopRequireDefault(_cliTable);

var _async = require('../../utils/async');

var _contentfulClients = require('../../utils/contentful-clients');

var _log = require('../../utils/log');

var _pagination = require('../../utils/pagination');

var _pagination2 = _interopRequireDefault(_pagination);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const command = exports.command = 'list';

const desc = exports.desc = 'List your organizations';

const builder = exports.builder = yargs => {
  return yargs.usage('Usage: contentful organization list').option('management-token', {
    alias: 'mt',
    describe: 'Contentful management API token',
    type: 'string'
  }).epilog(['See more at:', 'https://github.com/contentful/contentful-cli/tree/master/docs/organization/list', 'Copyright 2018 Contentful, this is a BETA release'].join('\n'));
};

const aliases = exports.aliases = ['ls'];

async function organizationList({ context }) {
  const { managementToken } = context;

  const client = await (0, _contentfulClients.createManagementClient)({
    accessToken: managementToken,
    feature: 'organization-list'
  });

  const result = await (0, _pagination2.default)({ client, method: 'getOrganizations' });

  const organizations = result.items.sort((a, b) => a.name.localeCompare(b.name));

  const table = new _cliTable2.default({
    head: ['Organization name', 'Organization id']
  });

  organizations.forEach(({ sys, name }) => {
    table.push([name, sys.id]);
  });

  (0, _log.log)(table.toString());
}

const handler = exports.handler = (0, _async.handleAsyncError)(organizationList);
