// contentful-migration/built/lib/intent-validator/content-type-update.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
const schema_validator_1 = __importDefault(require("./schema-validator"));
const Joi = __importStar(require("joi"));
class ContentTypeUpdateStepValidator extends schema_validator_1.default {
    constructor() {
        super(...arguments);
        this.article = 'a';
        this.displayName = 'content type';
    }
    appliesTo(step) {
        return step.isContentTypeUpdate();
    }
    get schema() {
        return {
            name: Joi.string().required(),
            description: Joi.string().required().allow(''),
            displayField: Joi.string().required()
        };
    }
}
exports.default = ContentTypeUpdateStepValidator;
