// contentful-migration/built/lib/intent-validator/field-update.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
const schema_validator_1 = __importDefault(require("./schema-validator"));
const Joi = __importStar(require("joi"));
class FieldUpdateStepValidator extends schema_validator_1.default {
    constructor() {
        super(...arguments);
        this.article = 'a';
        this.displayName = 'field';
    }
    appliesTo(step) {
        return step.isFieldUpdate();
    }
    get schema() {
        return {
            id: Joi.string().required(),
            newId: Joi.string().required(),
            name: Joi.string().required(),
            type: Joi.string().required(),
            localized: Joi.boolean().required().strict(),
            required: Joi.boolean().required().strict(),
            validations: Joi.array().required(),
            disabled: Joi.boolean().required().strict(),
            omitted: Joi.boolean().required().strict(),
            deleted: Joi.boolean().required().strict(),
            linkType: Joi.string().required(),
            items: Joi.object().required()
        };
    }
}
exports.default = FieldUpdateStepValidator;
